### ISSUE TITLE
Infer type of `var` variable from right-hand-side of assignment

### ISSUE THREAD

**Issue by xenoterracide** [typetools/checker-framework#4593]
```java
  Map<String, @Nullable Object> getContext( Map<String, String> context ) {
    var map = new HashMap<String, @Nullable Object>();
    map.putAll( context );
    map.putAll( mapper.convertValue( cliConfiguration, Map.class ) );
    map.put( "configDir", configDir.getParent() );
    return Collections.unmodifiableMap( map );
  }
```
fails to compile with 2 errors

```
home/xeno/IdeaProjects/brix/config-loader/spi/src/main/java/com/xenoterracide/brix/configloader/spi/ConfigValueProcessor.java:75: error: [assignment.type.incompatible] incompatible types in assignment.
    var map = new HashMap<String, @Nullable Object>();
              ^
  found   : @Initialized @NonNull HashMap<@Initialized @NonNull String, @Initialized @Nullable Object>
  required: @UnknownInitialization @Nullable HashMap<@Initialized @NonNull String, @Initialized @NonNull Object>
/home/xeno/IdeaProjects/brix/config-loader/spi/src/main/java/com/xenoterracide/brix/configloader/spi/ConfigValueProcessor.java:78: error: [argument.type.incompatible] incompatible argument for parameter value of put.
    map.put( "configDir", configDir.getParent() );
                                             ^
  found   : @Initialized @Nullable Path
  required: @Initialized @NonNull Object
```

but this code has no errors

```java
  Map<String, @Nullable Object> getContext( Map<String, String> context ) {
    Map<String, @Nullable Object> map = new HashMap<>();
    map.putAll( context );
    map.putAll( mapper.convertValue( cliConfiguration, Map.class ) );
    map.put( "configDir", configDir.getParent() );
    return Collections.unmodifiableMap( map );
  }
  ```

the only difference is that I'm not using `var` (I believe var uses left hand side inference as opposed to this being right hand side, but I may have them inverted)

related to https://github.com/typetools/checker-framework/issues/4587#issue-870243279 but that's a request for better documentation, this is for an actual compiler improvement.

**Comment by mernst** [comment]
I added a simplified test case as `Issue4593.java`.

**Comment by spacether** [comment]
The var feature [JEP286](https://openjdk.org/jeps/286) was released in Java 10, which was released in 2018.
So this feature has been around for 6 years.
I am still seeing this issue in the checker framework v3.34.0
Any chance that fixing this can be prioritized? Thank you for all of your hard work!

**Comment by msridhar** [comment]
@spacether I believe this was fixed by https://github.com/typetools/checker-framework/pull/5770, which was included in v3.35.0.  There have been further fixes since then, so I suggest you give things a try with the latest Checker Framework release.

I'm going to go ahead and close this issue (sorry we missed closing it before), but if you notice a problem with `var` support please file a new issue.

**Comment by spacether** [comment]
Upgrading fixed it, thanks!
